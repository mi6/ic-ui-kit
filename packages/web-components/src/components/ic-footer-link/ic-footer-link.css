@import "../../global/normalize.css";

:host(.footer-link-ungrouped-sparse) {
  --footer-link-margin-left: 0;
  --footer-link-margin-right: 2.5rem;
  --footer-link-margin-bottom: 0;
}

:host(.footer-link-ungrouped-small) {
  --footer-link-margin-left: 0;
  --footer-link-margin-right: 0;
  --footer-link-margin-bottom: var(--ic-space-md);
}

:host(.footer-link-grouped-sparse) {
  --footer-link-margin-left: 0;
  --footer-link-margin-right: 0;
  --footer-link-margin-bottom: var(--ic-space-md);
}

:host(.footer-link-grouped-small) {
  --footer-link-margin-left: var(--ic-space-md);
  --footer-link-margin-right: 0;
  --footer-link-margin-bottom: var(--ic-space-md);
}

:host(.footer-link) {
  margin-left: var(--footer-link-margin-left);
  margin-right: var(--footer-link-margin-right);
  margin-bottom: var(--footer-link-margin-bottom);
}

/* Link styling */

:host(.footer-link) a,
:host(.footer-link) a ::slotted(a) {
  display: flex;
  align-items: flex-end;
  font-weight: var(--ic-font-weight-bold);
  color: inherit;
  width: fit-content;
  font-size: var(--ic-font-size-regular);
}

:host(.footer-link) a:link,
:host(.footer-link) a ::slotted(a:link) {
  color: inherit;
  text-decoration: underline;
  font-weight: var(--ic-font-weight-bold);
}

:host(.footer-link) a:link > ::slotted(svg),
:host(.footer-link) a:link > ::slotted(img),
:host(.footer-link) a ::slotted(a:link > ::slotted(svg)),
:host(.footer-link) a ::slotted(a:link > ::slotted(img)) {
  padding: var(--ic-space-xxxs);
  pointer-events: none;
  transition: var(--ic-easing-transition-fast);
}

/*
There's an apparent defect in Chrome that means this isn't always re-rendered if changed.
Not an issue in Safari. Line 73-79 implement the same function but less clean.

a:link > ::slotted(svg) ,
a:link:visited > ::slotted(svg) {
  fill: var(--ic-theme-text);
}
*/

:host(.footer-link-light) ::slotted(svg) {
  fill: var(--ic-architectural-white);
}

:host(.footer-link-dark) ::slotted(svg) {
  fill: var(--ic-architectural-black);
}

:host(.footer-link) a:link:hover,
:host(.footer-link) a:link:focus,
:host(.footer-link) a ::slotted(a:link:hover),
:host(.footer-link) a ::slotted(a:link:focus) {
  border-bottom: 0.25rem solid !important;
  margin-bottom: -0.25rem !important;
  text-decoration: none;
}

@supports (text-decoration-thickness: 25%) {
  :host(.footer-link) a:link:hover,
  :host(.footer-link) a:link:focus,
  :host(.footer-link) a ::slotted(a:link:hover),
  :host(.footer-link) a ::slotted(a:link:focus) {
    text-decoration-line: underline;
    text-decoration-thickness: 25%;
    text-underline-offset: 25%;
    border-bottom: 0 !important;
    margin-bottom: 0 !important;
  }
}

:host(.footer-link) a:link:hover,
:host(.footer-link) a ::slotted(a:link:hover) {
  outline: none;
}

:host(.footer-link) a:link:focus,
:host(.footer-link) a ::slotted(a:link:focus) {
  outline: var(--ic-hc-focus-outline);
}

:host(.footer-link) a:link:focus > ::slotted(svg),
:host(.footer-link) a:link:focus > ::slotted(img),
:host(.footer-link) a ::slotted(a:link:focus > ::slotted(svg)),
:host(.footer-link) a ::slotted(a:link:focus > ::slotted(img)) {
  box-shadow: var(--ic-border-focus);
  border-radius: var(--ic-border-radius);
}

:host(.footer-link) a:focus,
:host(.footer-link) a ::slotted(a:focus) {
  border-radius: var(--ic-border-radius);
  outline: var(--ic-hc-focus-outline);
  transition: var(--ic-transition-duration-fast);
}

:host(.footer-link) a:link:active,
:host(.footer-link) a:link:focus:active,
:host(.footer-link) a:link:visited:active,
:host(.footer-link) a ::slotted(a:link:active),
:host(.footer-link) a ::slotted(a:link:focus:active),
:host(.footer-link) a ::slotted(a:link:visited:active) {
  color: inherit;
  text-decoration: none;
}

:host(.footer-link) a:hover,
:host(.footer-link) a ::slotted(a:hover) {
  text-decoration: none;
}

.open-in-new-icon {
  padding-left: var(--ic-space-xxs);
}

.open-in-new-icon,
.open-in-new-icon svg {
  height: var(--ic-space-md);
  width: var(--ic-space-md);
}

:host(.footer-link-light) .open-in-new-icon svg {
  fill: var(--ic-architectural-white);
}

:host(.footer-link-dark) .open-in-new-icon svg {
  fill: var(--ic-architectural-black);
}

@media screen and (min-width: 769px) {
  :host(.footer-link.footer-link-ungrouped-sparse:last-of-type) {
    margin-right: 0;
  }
}

@media (forced-colors: active) {
  :host(.footer-link-light) ::slotted(svg),
  :host(.footer-link-dark) ::slotted(svg) {
    fill: currentcolor;
  }
}
