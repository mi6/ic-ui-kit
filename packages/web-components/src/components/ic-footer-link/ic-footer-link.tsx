import {
  Component,
  Element,
  Host,
  Listen,
  Prop,
  h,
  State,
} from "@stencil/core";
import { DEVICE_SIZES, getThemeForegroundColor } from "../../utils/helpers";
import { IcTheme, IcThemeForeground } from "../../utils/types";

import OpenInNew from "../../assets/OpenInNew.svg";

type FooterConfig = { small: boolean; grouped: boolean };

@Component({
  tag: "ic-footer-link",
  styleUrl: "ic-footer-link.css",
  shadow: {
    delegatesFocus: true,
  },
})
export class FooterLink {
  @Element() el: HTMLIcFooterLinkElement;

  @State() deviceSize: number = DEVICE_SIZES.XL;
  @State() footerConfig: FooterConfig = { small: false, grouped: false };
  @State() foregroundColor: IcThemeForeground = getThemeForegroundColor();

  /**
   * If `true`, the user can save the linked URL instead of navigating to it.
   */
  @Prop() download?: string | boolean = false;

  /**
   * The URL that the link points to.
   */
  @Prop() href?: string = null;

  /**
   * The human language of the linked URL.
   */
  @Prop() hreflang?: string;

  /**
   * How much of the referrer to send when following the link.
   */
  @Prop() referrerpolicy?: ReferrerPolicy;

  /**
   * The relationship of the linked URL as space-separated link types.
   */
  @Prop() rel?: string;

  /**
   * The place to display the linked URL, as the name for a browsing context (a tab, window, or iframe).
   */
  @Prop() target?: string;

  componentWillLoad(): void {
    this.footerConfig = this.inferConfig(this.el);
  }

  @Listen("footerResized", { target: "document" })
  footerResizeHandler(): void {
    this.footerConfig = this.inferConfig(this.el);
  }

  @Listen("themeChange", { target: "document" })
  footerThemeChangeHandler(ev: CustomEvent): void {
    const theme: IcTheme = ev.detail;
    this.foregroundColor = theme.mode;
  }

  private inferConfig(e: HTMLElement): FooterConfig {
    if (e.parentElement !== null) {
      if (e.parentElement.classList.contains("footer")) {
        return {
          small: e.parentElement.classList.contains("footer-small"),
          grouped: e.parentElement.classList.contains("footer-grouped"),
        };
      } else {
        return this.inferConfig(e.parentElement);
      }
    } else {
      return { small: false, grouped: false };
    }
  }

  render() {
    const {
      footerConfig,
      href,
      hreflang,
      referrerpolicy,
      rel,
      target,
      download,
    } = this;
    const { small, grouped } = footerConfig;

    return (
      <Host
        class={{
          "footer-link": true,
          [`footer-link-${grouped ? "grouped" : "ungrouped"}-${
            small ? "small" : "sparse"
          }`]: true,
          [`footer-link-${this.foregroundColor}`]: true,
        }}
      >
        <a
          href={href}
          hreflang={hreflang}
          referrerPolicy={referrerpolicy}
          rel={rel}
          download={download !== false ? download : null}
          target={target}
        >
          <slot />
          {target === "_blank" && (
            <span class="open-in-new-icon" innerHTML={OpenInNew} />
          )}
        </a>
      </Host>
    );
  }
}
