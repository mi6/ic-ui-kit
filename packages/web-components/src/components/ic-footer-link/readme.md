# ic-footer-link



<!-- Auto Generated Below -->


## Properties

| Property         | Attribute        | Description                                                                                         | Type                                                                                                                                                                                     | Default     |
| ---------------- | ---------------- | --------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| `download`       | `download`       | If `true`, the user can save the linked URL instead of navigating to it.                            | `boolean \| string`                                                                                                                                                                      | `false`     |
| `href`           | `href`           | The URL that the link points to.                                                                    | `string`                                                                                                                                                                                 | `null`      |
| `hreflang`       | `hreflang`       | The human language of the linked URL.                                                               | `string`                                                                                                                                                                                 | `undefined` |
| `referrerpolicy` | `referrerpolicy` | How much of the referrer to send when following the link.                                           | `"" \| "no-referrer" \| "no-referrer-when-downgrade" \| "origin" \| "origin-when-cross-origin" \| "same-origin" \| "strict-origin" \| "strict-origin-when-cross-origin" \| "unsafe-url"` | `undefined` |
| `rel`            | `rel`            | The relationship of the linked URL as space-separated link types.                                   | `string`                                                                                                                                                                                 | `undefined` |
| `target`         | `target`         | The place to display the linked URL, as the name for a browsing context (a tab, window, or iframe). | `string`                                                                                                                                                                                 | `undefined` |


----------------------------------------------


