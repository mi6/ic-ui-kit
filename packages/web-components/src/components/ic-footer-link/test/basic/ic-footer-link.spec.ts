import { newSpecPage } from "@stencil/core/testing";
import { FooterLink } from "../../ic-footer-link";

describe("ic-footer-link", () => {
  it("should render", async () => {
    const page = await newSpecPage({
      components: [FooterLink],
      html: `<ic-footer-link href="/">Link</ic-footer-link>`,
    });

    expect(page.root).toEqualHtml(`
    <ic-footer-link class="footer-link footer-link-light footer-link-ungrouped-sparse" href="/">
        <mock:shadow-root>
            <a href="/">
                <slot></slot>
            </a>
        </mock:shadow-root>
            Link
    </ic-footer-link>`);
  });

  it("should render small with grouped links", async () => {
    const page = await newSpecPage({
      components: [FooterLink],
      html: `<ic-footer-link href="/">Link</ic-footer-link>`,
    });

    page.rootInstance.footerConfig = { small: true, grouped: true };
    await page.waitForChanges();

    expect(page.root).toEqualHtml(`
        <ic-footer-link class="footer-link footer-link-light footer-link-grouped-small" href="/">
            <mock:shadow-root>
                <a href="/">
                    <slot></slot>
                </a>
            </mock:shadow-root>
                Link
        </ic-footer-link>`);
  });

  it("should set foregroundColor on theme change", async () => {
    const page = await newSpecPage({
      components: [FooterLink],
      html: `<ic-footer-link label="button1" onclick="alert('test')">
      </ic-footer-link>`,
    });

    await page.rootInstance.footerThemeChangeHandler({
      detail: { mode: "light" },
    });
    await page.waitForChanges();

    expect(page.rootInstance.foregroundColor).toEqual("light");
  });

  it("should test footer resize handler", async () => {
    const page = await newSpecPage({
      components: [FooterLink],
      html: `<ic-footer-link label="button1" onclick="alert('test')">
      </ic-footer-link>`,
    });

    await page.rootInstance.footerResizeHandler();
    await page.waitForChanges();

    expect(page.rootInstance.footerConfig).toStrictEqual({
      grouped: false,
      small: false,
    });
  });
});
