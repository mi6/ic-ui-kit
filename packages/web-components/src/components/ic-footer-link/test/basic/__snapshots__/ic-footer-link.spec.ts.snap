// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-footer-link should render small with grouped links: small-footer-link-with-grouped-links 1`] = `
<ic-footer-link class="footer-link footer-link-grouped-small footer-link-light" href="/" role="listitem">
  <mock:shadow-root>
    <ic-link href="/">
      <slot></slot>
    </ic-link>
  </mock:shadow-root>
  Link
</ic-footer-link>
`;

exports[`ic-footer-link should render: footer-link 1`] = `
<ic-footer-link class="footer-link footer-link-light footer-link-ungrouped-sparse" href="/" role="listitem">
  <mock:shadow-root>
    <ic-link href="/">
      <slot></slot>
    </ic-link>
  </mock:shadow-root>
  Link
</ic-footer-link>
`;
